add collaboration blog with gdotv (#29)
localstack/localstack-blog.github.io: content/2024-04-11-localstack-neptune-development-with-gv-gremlin-ide/index.md

images: []
contributors: ["Arthur Bigeard", "Harsh Mishra"]
tags: ['showcase']
---

## Introduction

LocalStack offers a rich cloud development framework to emulate the functionality of many AWS services locally at a fraction of the cost and effort. It supports developers with facilities to build, test and launch complex applications locally, including services such as the Amazon Neptune Graph database. 

However, interacting with Amazon Neptune can be challenging due to the lack of easy-to-use tools, as well as the learning curve associated with graph databases and their querying languages. We’re excited to announce our partnership with G.V() to bring you a first-class graph database development experience fully compatible with LocalStack and Amazon Neptune.

## What is G.V()?

G.V() is a Gremlin Integrated Development Environment (IDE for short) specially designed for Apache TinkerPop Graph Databases systems such as Amazon Neptune, Aerospike Graph and JanusGraph.

It’s an easy-to-install desktop application compatible with Windows, macOS and Linux. It provides many features to assist with day-to-day development and production usage of graph databases, including:

> -   Querying the graph database using Gremlin via a powerful smart editor offering code completion feature and embedded documentation for the Gremlin language and your data model.

Reviewer: ```suggestion
-   Querying the graph database using Gremlin via a powerful smart editor offering code completion and embedded documentation for the Gremlin language and your data model.
```

> -   Automatic discovery of your Graph data schema with visualisation and documentation features.

Reviewer: nit: "visualisation" -> Do we want to adopt British spelling in this blog? (just to confirm, as we usually use American, but should be fine as long it is consistent within a single document.)

With G.V(), developers and data analysts can drastically accelerate their graph database adoption and analyse their existing graph data effortlessly thanks to its data explorer, AI integration, graph visualisation capabilities and smart query editor.

## Configuring a LocalStack Neptune database connection on G.V()

A database connection in G.V() allows you to execute Gremlin queries, view your graph data schema, explore your data interactively and much more. To get started you’ll first need to install G.V(). You’ll also need to have LocalStack running with at least one Neptune instance created and its port number, as documented [here](https://docs.localstack.cloud/user-guide/aws/neptune/#getting-started).

### Installing G.V()

To install G.V(), head over to the [G.V() website](https://gdotv.com) and download the right executable for your operating system (Windows/macOS/Linux). Run the installer and follow the on -screen instructions, and you’ll be ready in just a few clicks.

### Connecting G.V() to your LocalStack Neptune Database

On G.V(), click on Add New Database Connection. Select LocalStack as the Graph Technology Type.

> Enter localhost.localstack.cloud as the hostname and the port number returned when creating your Neptune instance using the AWS CLI, then click on Test Connection. G.V() will make sure it can connect to your LocalStack Neptune Database as expected.

Reviewer: ```suggestion
Enter `localhost.localstack.cloud` as the hostname and the port number returned when creating your Neptune instance using the AWS CLI, then click on Test Connection. G.V() will make sure it can connect to your LocalStack Neptune Database as expected.
```

  property(single, 'age', 27).
  property(single, 'name', "vadas").
  addE('created').
    from('38a15c92-5e45-4d13-9c96-1dc4582e223d').
    to('b70c011a-6908-4ba6-ba09-3cf497a7200e').
  aggregate('t').
  property('weight', 0.4).
  addE('created').
    from('83b6b6eb-6dd5-43cd-9868-9b2dca746011').
    to('b70c011a-6908-4ba6-ba09-3cf497a7200e').
  aggregate('t').
  property('weight', 0.2).
  addE('created').
    from('aee846b6-c69b-4453-83ef-3d418cc84788').
    to('1b42b83e-a0df-4687-a92b-22defdf4ae30').
  aggregate('t').
  property('weight', 1).
  addE('knows').
    from('38a15c92-5e45-4d13-9c96-1dc4582e223d').
    to('603fe545-1b65-4ece-9f4a-e66005a30c3c').
  aggregate('t').
  property('weight', 0.5).
  addE('created').
    from('aee846b6-c69b-4453-83ef-3d418cc84788').
    to('b70c011a-6908-4ba6-ba09-3cf497a7200e').
  aggregate('t').
  property('weight', 0.4).
  addE('knows').
    from('38a15c92-5e45-4d13-9c96-1dc4582e223d').
    to('aee846b6-c69b-4453-83ef-3d418cc84788').
  aggregate('t').
  property('weight', 1).
  select('t')
```

Run the query by clicking on the green play button, and you should see the graph output shown below:

> {{< img-simple src="graph-output-gdotv" alt="Graph output in gdotv" width="800">}}

Reviewer: ```suggestion
{{< img-simple src="graph-output-gdotv.png" alt="Graph output in gdotv" width="800">}}
```

In the Connections sidebar, under your LocalStack connection, simply click on View Graph Data Model.

For the above sample data you’ll see the graph data schema below - it’s all generated automatically by G.V():

{{< img-simple src="gdotv-graph-data-schema.png" alt="Graph data schema in gdotv" width="800">}}

### Exploring the Graph Data interactively

You don’t always need to write a Gremlin query to explore your data - you can simply use G.V()’s Graph Data Explorer. You can design your query with a few quick filters and get straight to visualising your data.

{{< img-simple src="exploring-graph-data-interactively.png" alt="Exploring the Graph Data interactively" width="800">}}

There are many additional features directly in the Graph view. You can filter and navigate data interactively either using the search bar or the smart filter tool. You can also edit vertices and edges directly through the user interface, click on an element to view its properties and details, or edit them.

> We offer a wealth of customisation options to help you create unique visuals for your data. Make sure to explore all of those as they will be especially useful when reporting on data in your production database, for instance.

Reviewer 2: ```suggestion
G.V() offers a wealth of customisation options to help you create unique visuals for your data. Make sure to explore all of those as they will be especially useful when reporting on data in your production database, for instance.
```
It sounds a bit confusing that it suddenly says "We offer ..." when before the article sounded like a guide on how to use `G.V()`. This way people may assume this another LS product? 🤔

## Conclusion

With G.V(), you can considerably enhance your graph database development experience whilst gaining access to a powerful reporting and visualisation toolset for your production data. We’re just scratching the surface on what G.V() can do, and you can find out more on their documentation and their blog. Thanks to LocalStack’s core cloud emulator, parity is ensured between a local Neptune instance and AWS’s own, meaning Gremlin queries in your development environment will behave identically on Amazon Neptune.

> If you have questions about configuring and running your project, you can find the G.V() team on the official [Apache TinkerPop Discord](https://discord.gg/tinkerpop) or the on LocalStack team on our [Slack Community](https://localstack.cloud/slack)!

Reviewer: ```suggestion
If you have questions about configuring and running your project, you can find the G.V() team on the official [Apache TinkerPop Discord](https://discord.gg/tinkerpop) or the LocalStack team on our [Slack Community](https://localstack.cloud/slack)!
```
